### ISSUE TITLE
Implement support for containerd as CRI

### ISSUE THREAD

**Issue by vpnachev** [gardener/gardener-extension-os-suse-chost#42]
**How to categorize this issue?**

/area os usability
/kind enhancement
/priority 3
/os suse-chost

**What would you like to be added**:
Implement the requirements from https://github.com/gardener/gardener/blob/master/docs/proposals/10-shoot-additional-container-runtimes.md#design-details.

**Why is this needed**:
To enable containerd as CRI for CoreOS/Flatcar. Also needed for the docker deprecation https://github.com/gardener/gardener/issues/4110

**Comment by vpnachev** [comment]
Some findings so far
1. The binaries are installed in `/usr/sbin` and need to be linked from `/usr/bin`.
2. `ctr` is available as `containerd-ctr`, so it also have to be renamed/linked  as `ctr`.
3. `containerd.service` file is missing in the OS, so it cannot be simply started with `systemctl start containerd`
4. `containerd-shim-runc-v[12]` are missing

**Comment by vpnachev** [comment]
Some other findings, `runc` is also missing on the node
- I tried to re-use `docker-runc` by symbolic link
- Alternatively, installed `runc` from https://github.com/opencontainers/runc/releases/download/v1.0.0-rc95/runc.amd64

In both cases the nodes are healthy and pods can be created, but pod termination is failing with
```
error killing pod: [failed to "KillContainer" for "webserver" with KillContainerError: "rpc error: code = Unknown desc = failed to stop container \"f46c66da7b7118b0207cc0b5a93c7f3c119e1f842fe9acf40498c4057327527e\": unknown error after kill: runc did not terminate sucessfully: container_linux.go:419: signaling init process caused: permission denied\n: unknown", failed to "KillPodSandbox" for "9df16fc6-243e-4259-a24e-39d95b5ea71d" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to stop container \"f46c66da7b7118b0207cc0b5a93c7f3c119e1f842fe9acf40498c4057327527e\": failed to kill container \"f46c66da7b7118b0207cc0b5a93c7f3c119e1f842fe9acf40498c4057327527e\": unknown error after kill: runc did not terminate sucessfully: container_linux.go:419: signaling init process caused: permission denied\n: unknown"]
```
Because of this, the conformance tests are failing, so this setup cannot be used in productive clusters.



Here is the shoot spec that I am using to run the validation
```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: foo
  namespace: bar
spec:
  addons:
    kubernetesDashboard:
      authenticationMode: basic
      enabled: false
    nginxIngress:
      enabled: false
      externalTrafficPolicy: Cluster
  cloudProfileName: aws
  hibernation:
    enabled: false
  kubernetes:
    allowPrivilegedContainers: false
    kubeAPIServer:
      enableAnonymousAuthentication: false
      enableBasicAuthentication: false
      requests:
        maxMutatingInflight: 200
        maxNonMutatingInflight: 400
    kubeControllerManager:
      nodeCIDRMaskSize: 24
      nodeMonitorGracePeriod: 2m0s
      podEvictionTimeout: 2m0s
    kubeProxy:
      enabled: true
      mode: IPTables
    kubelet:
      failSwapOn: true
      kubeReserved:
        cpu: 80m
        memory: 1Gi
        pid: 20k
    version: 1.20.6
  maintenance:
    autoUpdate:
      kubernetesVersion: false
      machineImageVersion: false
    timeWindow:
      begin: 010000+0000
      end: 020000+0000
  networking:
    nodes: 10.250.0.0/16
    pods: 100.96.0.0/11
    services: 100.64.0.0/13
    type: calico
  provider:
    infrastructureConfig:
      apiVersion: aws.provider.extensions.gardener.cloud/v1alpha1
      kind: InfrastructureConfig
      networks:
        vpc:
          cidr: 10.250.0.0/16
        zones:
        - internal: 10.250.112.0/22
          name: eu-west-1a
          public: 10.250.96.0/22
          workers: 10.250.0.0/19
    type: aws
    workers:
    - cri:
        name: containerd
      machine:
        image:
          name: suse-chost
          version: 15.2.20210610 # ami-0cae8a3a77df68846
        type: c5.large
      maxSurge: 1
      maxUnavailable: 0
      maximum: 2
      minimum: 2
      name: containerd
      systemComponents:
        allow: true
      volume:
        size: 50Gi
        type: gp2
      zones:
      - eu-west-1a
  purpose: development
  region: eu-west-1
```

In this branch https://github.com/vpnachev/gardener-extension-os-suse-chost/commits/containerd I am pushing the changes that I am currently using to enable containerd in this extension.

**Comment by vpnachev** [comment]
I have tried this also with SP3 image `ami-05afe81ca301503b1`. 
This time the runc and containerd-shim-runc-v[12] binaries are available and it is no hacks with symbolic links are required.
The containerd.service is still missing, but this is not a big deal for now. 

Unfortunately, the above error with `permission denied` is still present.

**Comment by vpnachev** [comment]
The permission error seems to be related to the AppArmor. It is enabled on the suse node
```bash
$ cat /sys/module/apparmor/parameters/enabled
Y
$ cat /sys/module/apparmor/parameters/mode 
enforce
```

And the only profile that is enabled is 
```bash
$ cat /sys/kernel/security/apparmor/profiles
cri-containerd.apparmor.d (enforce)
```

but I guess it is not sufficient to run any kind of pod/container and runc often fail due to insufficient permissions. 

When I run something like `watch aa-teardown` on the node, I manage to let container runtime terminate and delete the containers, then the kubelets can finally complete the pod deletion and I no longer see the error 

```
error killing pod: [failed to "KillContainer" for "webserver" with KillContainerError: "rpc error: code = Unknown desc = failed to stop container \"f46c66da7b7118b0207cc0b5a93c7f3c119e1f842fe9acf40498c4057327527e\": unknown error after kill: runc did not terminate sucessfully: container_linux.go:419: signaling init process caused: permission denied\n: unknown", failed to "KillPodSandbox" for "9df16fc6-243e-4259-a24e-39d95b5ea71d" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to stop container \"f46c66da7b7118b0207cc0b5a93c7f3c119e1f842fe9acf40498c4057327527e\": failed to kill container \"f46c66da7b7118b0207cc0b5a93c7f3c119e1f842fe9acf40498c4057327527e\": unknown error after kill: runc did not terminate sucessfully: container_linux.go:419: signaling init process caused: permission denied\n: unknown"]
```

However, might not be sufficient as I still got some 20 failed test cases out of 305 executed, but I guess I need proper AppArmor profiles or to properly disable it and re-run the tests again as this `watch aa-teardown` is not reliable mitigation.

**Comment by dcermak** [comment]
> And the only profile that is enabled is
> 
> ```shell
> $ cat /sys/kernel/security/apparmor/profiles
> cri-containerd.apparmor.d (enforce)
> docker-default (enforce)
> ```

Is docker running on the host as well?

> When I run something like `watch aa-teardown` on the node, I manage to let container runtime terminate and delete the containers, then the kubelets can finally complete the pod deletion and I no longer see the error

I assume you can reproduce this, right? Could you try to trigger the issue again and post the contents of `/var/log/apparmor/` and `/var/log/audit.log*`? I guess that something with the generated apparmor profile is wrong.

**Comment by vpnachev** [comment]
> Is docker running on the host as well?

Yes, it is.

> I assume you can reproduce this, right? Could you try to trigger the issue again and post the contents of /var/log/apparmor/ and /var/log/audit.log*? I guess that something with the generated apparmor profile is wrong.

Yes, it is reproducible. I am running k8s conformance tests via our test-infra, [ref](https://github.com/gardener/test-infra/blob/master/integration-tests/e2e/README.md#example-usage-conformance-tests). This week I am busy with some other tasks, but will take a look next week and share the logs then.

**Comment by vpnachev** [comment]
Just for reference, I am no longer installing containerd-shim and runc myself but using those that are now available on the node.

Also, removed the symbolic link `/usr/bin/containerd -> /usr/sbin/containerd` and let the systemd service use `/usr/sbin/containerd` just to make sure that this is not causing troubles to the apparmor profile.

```bash
$ runc -v
runc version 1.0.0-rc93
commit: 12644e614e25b05da6fd08a38ffa0cfe1903fdec
spec: 1.0.2-dev
go: go1.13.15
libseccomp: 2.4.1

$ containerd -v
containerd github.com/containerd/containerd v1.4.4 05f951a3781f4f2c1911b05e61c160e9c30eaa8e

**Comment by dcermak** [comment]
> Thanks for the feedback @dcermak !
> 
> TL;DR - I will have to sort out with my colleagues how to install the additional packages as by default this is not possible.

I can build you a VM with the required packages in your desired format in the build service. Just give me the package list and your desired format.


> _Long version_: Just to try it out, I have configured openSuSE Leap repos and installed `apparmor-parser`, then the profile couldn't be loaded because `tunables/global` are missing, so I installed even further packages from OpenSuSE (the entire `apparmor` pattern). At the end I have loaded the profile, but there is plenty of other that have been loaded too. Some packages have been re-installed or downgraded. And currently I believe I have altered too much the system that the results of the validations might be wrong, therefore I want to do it with the SuSE packages.

I don't have a perfect solution for this, but what you can do is this:
1. add the free SLE BCI repository on your host (this is a free subset of the full SLES repositories that is free to use without a subscription):
```ShellSession
$ cat /etc/zypp/repos.d/SLE_BCI.repo
[SLE_BCI]
enabled=1
autorefresh=0
baseurl=https://updates.suse.com/SUSE/Products/SLE-BCI/$releasever_major-SP$releasever_minor/$basearch/product/
priority=100
```
2. Install `apparmor-parser` from that.
3. omit `tunables/global` from the apparmor profile that I shared, I have included it only for completenes, but it shouldn't be required 🤞
